google-cloud-firestore/lib/google/cloud/firestore/bulk_writer.rb

```diff
@@ -0,0 +1,458 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License")
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "algorithms"
+require "concurrent"
+require "google/cloud/firestore/rate_limiter"
+require "google/cloud/firestore/bulk_commit_batch"
+require "google/cloud/firestore/concurrent/promises"
+require "google/cloud/firestore/bulk_writer_operation"
+require "google/cloud/firestore/bulk_writer_exception"
+require "google/cloud/firestore/bulk_writer_scheduler"
+
+
+module Google
+  module Cloud
+    module Firestore
+      class BulkWriter
+        MAX_RETRY_ATTEMPTS = 15
+
+        ##
+        # Initialize the attributes and start the schedule_operations job
+        #
+        def initialize client, service,
+                       request_threads: nil,
+                       batch_threads: nil,
+                       retries: MAX_RETRY_ATTEMPTS
+          @client = client
+          @service = service
+          @closed = false
+          @flush = false
+          @request_threads = (request_threads || 2).to_i
+          @write_thread_pool = Concurrent::ThreadPoolExecutor.new max_threads: @request_threads,
+                                                                  max_queue: 0
+          @mutex = Mutex.new
+          @scheduler = BulkWriterScheduler.new client, service, batch_threads
+          @doc_refs = Set.new
+          @retries = [retries || MAX_RETRY_ATTEMPTS, MAX_RETRY_ATTEMPTS].min
```

Comment:
This can just be `[retries,MAX_RETRY_ATTEMPTS].min` as `retries` is already defaulted to max value.

Reply:
From the client class nil value is being passed which overrides the default value in the method argument. Have updated the method signature instead.